### PR TITLE
GlobalContext: Make context configuration idempotent

### PR DESCRIPTION
**Pull request by jzila**
This is to allow KBFS to use a context separately from `libkb.G`, so that we can expand the `libkbfs.Context` interface without requiring that `*libkb.GlobalContext` implement all its methods.

### CODE REVIEW

**Comment by maxtaco** [comment]
Can't merge this one until after the mobile release for teams.

**Comment by jzila** [comment]
Does it break something?

**Comment by jzila** [comment]
Or is it a matter of risk?

**Comment by jzila** [comment]
For now I can vendor it into KBFS and merge into client/master later.

**Comment by maxtaco** [comment]
No risky changes in until after the apps are in review.  This one feels
risky since it changes code the runs on mobile and is hard to rest.  I will
review it tonight!

**Comment by maxtaco** [comment]
Test.

File: go/libkb/globals.go [hunk 1]
```diff
@@ -127,6 +127,13 @@ type GlobalContext struct {
 	ActiveDevice *ActiveDevice
 
 	NetContext context.Context
+
+	// Initialization variables for idempotency
+	isInitDone               bool
```
Comment (songgao) [inline go/libkb/globals.go]: If this is only done once and never reset back to `false`, can we use `sync.Once` here instead?

Comment (jzila) [reply go/libkb/globals.go]: I guess so, but some of them can error out, in which case it shouldn't get set. So some will be booleans and some will be `sync.Once`.

Comment (songgao) [reply go/libkb/globals.go]: I see. For booleans, do we need worry about locking?

Comment (jzila) [reply go/libkb/globals.go]: Yeah, we probably do.

Comment (jzila) [reply go/libkb/globals.go]: I'll make it all `sync.Once`, and then for the error ones, I'll use a bool + mutex.

Comment (songgao) [reply go/libkb/globals.go]: Sounds good; thanks!

**Comment by maxtaco** [comment]
I took a quick look; this should be OK, but it feels risky, because we might be depending upon buggy behavior as it is.  I need to take a deeper look, and I still think we need to hold off on the merge to master.

File: go/libkb/globals.go [hunk 2]
```diff
@@ -312,26 +322,38 @@ func (g *GlobalContext) Logout() error {
...
 	return nil
 }
 
 func (g *GlobalContext) PushShutdownHook(sh ShutdownHook) {
 	g.ShutdownHooks = append(g.ShutdownHooks, sh)
 }
 
-func (g *GlobalContext) ConfigureConfig() error {
-	err := RemoteSettingsRepairman(g)
+func (g *GlobalContext) ConfigureConfig() (err error) {
```
Comment (maxtaco) [inline go/libkb/globals.go]: I think in `systests` we need this to be called multiple files.

Comment (jzila) [reply go/libkb/globals.go]: Seems to work without it though.

**Comment by maxtaco** [comment]
I still don't understand why this is necessary.  Can you explain it to me in terms of the calling code?

File: go/libkb/globals.go (hunk 2 above)
Comment (maxtaco) [reply go/libkb/globals.go]: Can we let `ConfigureConfig` actually run twice if called twice?

**Comment by patrickxb** [comment]
I also don't understand why this is necessary.  What bug is it fixing?

Please, please do not merge/vendor into any repo until after teams are released.  Desktop/mobile, everything.

**Comment by patrickxb** [comment]
It's very early in the morning, but I'm worried about this in regards to Logout and switching users.

I will look in detail at office.

**Comment by patrickxb** [comment]
I'm pretty sure this could cause an issue in mobile bind that does its own G initialization.  No CI tests will test it.

Since this isn't fixing a bug, I think this PR should be abandoned.  This object is everywhere and I can't see any benefit.

Changes to GlobalContext should only be fixing critical bugs and need to be tested fully on all devices.

**Comment by maxtaco** [comment]
Yeah, first I'd like to understand why this is needed to see if there are better workarounds.

But if this is needed, seems possible to keep the current behavior as is and allow the requested behavior in a new codepath.

**Comment by jzila** [comment]
@maxtaco, @patrickxb: thanks for the feedback! I intended the changes for [this](https://github.com/keybase/kbfs/pull/1171/files#diff-43a873109b2675fa0d833cd81b7f68cfL25) diff, but I'm open to anything that achieves the same goal.

I need to create a `libkbfs.Context` from a passed-in context rather than the `libkb.G` static one. This is so `go-bind` [can create a KBFS context from a `*libkb.GlobalContext`](https://github.com/keybase/client/blob/86c8932fd58b8e06f03dfd5537f1977cc93be270/go/bind/keybase.go#L144-L152), whereas right now it's assumed that `*libkb.GlobalContext` implements `libkbfs.Context`. 

This PR is needed to remove the library-level `sync.Once` in KBFS that prevents reinitialization of the `*libkb.GlobalContext`. It's possible that I don't need to do that, instead calling the methods every time. This code was added in keybase/kbfs#52 to facilitate the mobile code, so I'm not sure what the right answer is.

cc @strib as well since he reviewed keybase/kbfs#52.

**Comment by jzila** [comment]
@maxtaco, @patrickxb: any suggestions for a better approach? This is blocking progress.

**Comment by maxtaco** [comment]
@jzila why doesn't it work to always run `NewContextFromGlobalContext` after the Global context is good to go?  I made this comment: https://github.com/keybase/kbfs/pull/1171#discussion_r138990510

**Comment by jzila** [comment]
I think the issue is that we don't know that each of those configure methods has already been run by the time we get to `NewContextFromGlobalContext`. `*GlobalContext` is plumbed everywhere, and I don't have a great grasp of the flow of how it gets to KBFS.

**Comment by patrickxb** [comment]
The one coming from `go/bind` is already configured.

**Comment by patrickxb** [comment]
Otherwise, it looks like you are just using the variable `libkb.G` (which is going away).  In that case, just make your own and configure it.

**Comment by maxtaco** [comment]
yeah i was going to suggest making a new one.

**Comment by maxtaco** [comment]
@jzila i made this: https://github.com/keybase/kbfs/tree/maxtaco/different-jzila-g

the branch as is didn't compile, but this at least compiles.

**Comment by maxtaco** [comment]
it shoudl work to just make your own G. if not, I will fix it.

**Comment by jzila** [comment]
@maxtaco: looks good, thanks! I'll take your commit as-is.

**Comment by maxtaco** [comment]
cool. if stuff breaks because it still wants a global G implicitly, I'll hunt down what's broken. thanks for your patience! sorry to be so paranoid, we've just been burned by changes here in the past.

**Comment by jzila** [comment]
Thanks @maxtaco and @patrickxb for the awesome help!

**Comment by maxtaco** [comment]
sure thing, glad that fix worked

**Pull request closed without merging**